### PR TITLE
18 bot offensive mode

### PR DESCRIPTION
**Pull request by chzlithoo**

### CODE REVIEW

File: playground/strategic_bot.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
+from collections import defaultdict
+from operator import itemgetter
+from typing import List, Dict
+
+from board import Stone, BoardConfig
+from util import turn_check, Direction
+
+
+# edge is left most or upper most stone in a connection
+def is_edge(s: Stone, log: List[Stone]):
```
Comment (hyoungseok) [inline playground/strategic_bot.py]: need return type

File: playground/strategic_bot.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        self.k = board_config.connect
+        self.m = board_config.row
+        self.n = board_config.column
+        self.p = board_config.each_move
+        self.q = board_config.first_move
+        # TODO: direction functions to be generalized to util (on the next phase)
+        self.directions = {
+            Direction.RIGHT: lambda x, y: (x + 1, y),
+            Direction.DOWN: lambda x, y: (x, y + 1),
+            Direction.DOWN_RIGHT: lambda x, y: (x + 1, y + 1),
```
Comment (hyoungseok) [inline playground/strategic_bot.py]: need positive diagonal case logic

File: playground/strategic_bot.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+			if m_stone in log:
+				directions.append(dd)
+		if len(directions) > 0:
+			is_edge = True
+		else:
+			is_edge = False
+		return is_edge, directions
+
+	# TODO: to be generalized to util (on the next phase)
+	def connection_check(self, edge: Edge, log: List[Stone], num: int):
```
Comment (hyoungseok) [inline playground/strategic_bot.py]: need return type

File: playground/strategic_bot.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+			s = self.optimal_stone(opponent, log)
+		else:
+			# mode offense
+			print("off")
+			s = self.optimal_stone(mine, log)
+		s.color = self.my_color
+		print(s)
+		return s
+
+	def filter_valid(self, corners: List[Stone], log: List[Stone]):
```
Comment (hyoungseok) [inline playground/strategic_bot.py]: need return type

File: playground/strategic_bot.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+			while True:
+				new_x, new_y = direction(int(new_x), int(new_y))
+				new_stone = Stone(str(new_x), str(new_y), edge.stone.color)
+				if new_stone in log:
+					connection_count += 1
+				else:
+					break
+			if connection_count == num:
+				is_connected = True
+				break
```
Comment (hyoungseok) [inline playground/strategic_bot.py]: could early return?

File: playground/strategic_bot.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+			is_edge, directions = self.is_stone_edge(s, log)
+			if s.color == color and is_edge:
+				edges[(s.x, s.y, s.color)] = directions
+		# returns connected that are not closed
+		for edge_, dirs in edges.items():
+			edge = Edge(Stone(edge_[0], edge_[1], edge_[2]), dirs)
+			for num in list(range(6, 1, -1)):
+				if self.connection_check(edge, log, num):
+					groups[num].append(edge)
+					break
```
Comment (hyoungseok) [inline playground/strategic_bot.py]: could early return?

File: playground/strategic_bot.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+			s = self.optimal_stone(mine, log)
+			s.color = self.my_color
+			return s
+		opponent_max_key = max(opponent_keys)
+		print("opponent max", opponent_max_key)
+		if opponent_max_key >= 3:
+			# mode defense
+			print("def")
+			s = self.optimal_stone(opponent, log)
+		else:
```
Comment (hyoungseok) [inline playground/strategic_bot.py]: could early return

File: playground/house.py [hunk 8]
```diff
@@ -7,85 +7,86 @@
 import util
 
 if not os.path.exists("templates"):
-    os.mkdir("templates")
+	os.mkdir("templates")
```
Comment (hyoungseok) [inline playground/house.py]: indent 6

File: playground/strategic_bot.py (hunk 6 above)
Comment (chzlithoo) [reply playground/strategic_bot.py]: can't return in current loop b/c it has double for loops

Comment (hyoungseok) [reply playground/strategic_bot.py]: got it, thanks

File: playground/strategic_bot.py (hunk 7 above)
Comment (chzlithoo) [reply playground/strategic_bot.py]: IMO, early return makes more code duplicates

File: playground/house.py (hunk 8 above)
Comment (chzlithoo) [reply playground/house.py]: I've already set my indent size as 4, plz help
image

**Comment by ths2o** [comment]
LGTM 👍

**Pull request merged**